src/tokenizer/tokenizer.rs

```diff
@@ -26,27 +28,72 @@ impl<T: Tokenizer> BoxableTokenizer for T {
     }
 }
 
-impl Clone for TextAnalyzer {
-    fn clone(&self) -> Self {
-        TextAnalyzer {
-            tokenizer: self.tokenizer.box_clone(),
-        }
+pub struct BoxedTokenizer(Box<dyn BoxableTokenizer>);
+
+impl Clone for BoxedTokenizer {
+    fn clone(&self) -> BoxedTokenizer {
+        Self(self.0.box_clone())
     }
 }
 
-impl Default for TextAnalyzer {
-    fn default() -> TextAnalyzer {
-        TextAnalyzer::from(EmptyTokenizer)
+impl Tokenizer for BoxedTokenizer {
+    type TokenStream<'a> = Box<dyn TokenStream + 'a>;
+
+    fn token_stream<'a>(&'a mut self, text: &'a str) -> Self::TokenStream<'a> {
+        self.0.box_token_stream(text).into()
     }
 }
 
-impl<T: Tokenizer + Clone> From<T> for TextAnalyzer {
-    fn from(tokenizer: T) -> Self {
-        TextAnalyzer::builder(tokenizer).build()
+/// Trait for the pluggable components of `Tokenizer`s.
+pub trait BoxableTokenFilter: 'static + Send + Sync {
+    /// Wraps a Tokenizer and returns a new one.
+    fn box_transform(&self, tokenizer: BoxedTokenizer) -> Box<dyn BoxableTokenizer>;
+}
+
+impl<T: TokenFilter> BoxableTokenFilter for T {
+    fn box_transform(&self, tokenizer: BoxedTokenizer) -> Box<dyn BoxableTokenizer> {
+        let tokenizer = self.clone().transform(tokenizer);
+        tokenizer.box_clone()
+    }
+}
+
+pub struct BoxTokenFilter(Box<dyn BoxableTokenFilter>);
+
+impl Deref for BoxTokenFilter {
+    type Target = dyn BoxableTokenFilter;
+
+    fn deref(&self) -> &dyn BoxableTokenFilter {
+        &*self.0
+    }
+}
+
+impl<T: TokenFilter> From<T> for BoxTokenFilter {
+    fn from(tokenizer: T) -> BoxTokenFilter {
+        BoxTokenFilter(Box::new(tokenizer))
     }
 }
 
 impl TextAnalyzer {
+    /// Builds a new `TextAnalyzer` given a tokenizer and a vector of `BoxTokenFilter`.
+    ///
+    /// When creating a `TextAnalyzer` from a `Tokenizer` alone, prefer using
+    /// `TextAnalyzer::from(tokenizer)`.
+    /// When creating a `TextAnalyzer` from a `Tokenizer` and a static set of `TokenFilter`,
+    /// prefer using `TextAnalyzer::builder(tokenizer).filter(token_filter).build()`.
```

Comment:
you should explain why it is preferred (i.e. less Boxes, better tokenization perf)